### PR TITLE
adding support for ec2 instance store volumes

### PR DESCRIPTION
**Pull request by saurabhpal97**
Fixes # 

**Description**

**How was this change tested?**

* make test

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bwagner5** [comment]
Hey! Apologies for the delay! I will try to get this reviewed tomorrow!

File: charts/karpenter/crds/karpenter.k8s.aws_awsnodetemplates.yaml [hunk 1]
```diff
@@ -121,6 +121,10 @@ spec:
                             in the Amazon Elastic Compute Cloud User Guide.
                           type: string
                       type: object
+                    virtualName:
+                      description: The device name for instance store volumes (for
+                        example, ephermal1)
```
Comment (bwagner5) [inline charts/karpenter/crds/karpenter.k8s.aws_awsnodetemplates.yaml:126]:
```suggestion
                        example, ephemeral1)
```

File: pkg/cloudprovider/aws/apis/v1alpha1/provider.go [hunk 2]
```diff
@@ -127,6 +127,8 @@ type MetadataOptions struct {
 type BlockDeviceMapping struct {
 	// The device name (for example, /dev/sdh or xvdh).
 	DeviceName *string `json:"deviceName,omitempty"`
+	// The device name for instance store volumes (for example, ephermal1)
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/apis/v1alpha1/provider.go:130]:
```suggestion
	// The device name for instance store volumes (for example, ephemeral1)
```

File: pkg/cloudprovider/aws/launchtemplate.go [hunk 3]
```diff
@@ -216,7 +216,8 @@ func (p *LaunchTemplateProvider) blockDeviceMappings(blockDeviceMappings []*v1al
 	blockDeviceMappingsRequest := []*ec2.LaunchTemplateBlockDeviceMappingRequest{}
 	for _, blockDeviceMapping := range blockDeviceMappings {
 		blockDeviceMappingsRequest = append(blockDeviceMappingsRequest, &ec2.LaunchTemplateBlockDeviceMappingRequest{
-			DeviceName: blockDeviceMapping.DeviceName,
+			DeviceName:  blockDeviceMapping.DeviceName,
+			VirtualName: blockDeviceMapping.VirtualName,
 			Ebs: &ec2.LaunchTemplateEbsBlockDeviceRequest{
 				DeleteOnTermination: blockDeviceMapping.EBS.DeleteOnTermination,
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/launchtemplate.go:222]: These should be empty when supplying VirtualName, so they will panic on nill dereference here.

**Comment by saurabhpal97** [comment]
Thanks for the comments @bwagner5 . I will surely work on the tests and suggested changes

**Comment by blakepettersson** [comment]
What's the status on this one? Do you still have time to work on this @saurabhpal97?

**Comment by saurabhpal97** [comment]
Sorry folks, been a bit occupied with work. I'll try to get this done over the weekend

**Comment by sftim** [comment]
If we merge this, I recommend also filing an issue about documenting the change.

File: pkg/cloudprovider/aws/launchtemplate.go (hunk 3 above)
Comment (saurabhpal97) [reply pkg/cloudprovider/aws/launchtemplate.go]: I didn't get this one, @bwagner5 . Can you help?

Comment (blakepettersson) [reply pkg/cloudprovider/aws/launchtemplate.go]: I think he means that `Ebs` should be set to `nil` if `VirtualName` is specified, so I guess something like

```go
		var ebs *ec2.LaunchTemplateEbsBlockDeviceRequest
		if blockDeviceMapping.VirtualName == nil {
			ebs = &ec2.LaunchTemplateEbsBlockDeviceRequest{
				DeleteOnTermination: blockDeviceMapping.EBS.DeleteOnTermination,
				Encrypted:           blockDeviceMapping.EBS.Encrypted,
				VolumeType:          blockDeviceMapping.EBS.VolumeType,
				Iops:                blockDeviceMapping.EBS.IOPS,
				Throughput:          blockDeviceMapping.EBS.Throughput,
				KmsKeyId:            blockDeviceMapping.EBS.KMSKeyID,
				SnapshotId:          blockDeviceMapping.EBS.SnapshotID,
				VolumeSize:          p.volumeSize(blockDeviceMapping.EBS.VolumeSize),
			}
		}

		blockDeviceMappingsRequest = append(blockDeviceMappingsRequest, &ec2.LaunchTemplateBlockDeviceMappingRequest{
			DeviceName:  blockDeviceMapping.DeviceName,
			VirtualName: blockDeviceMapping.VirtualName,
			Ebs:         ebs,
		})
```

**Comment by wkaczynski** [comment]
@bwagner5 would the ephemeral-storage capacity derivation mentioned by @jonathan-innis in https://github.com/aws/karpenter/issues/2723#issuecomment-1301371762 be a part of this PR or some follow-up work ?

**Comment by bwagner5** [comment]
> @bwagner5 would the ephemeral-storage capacity derivation mentioned by @jonathan-innis in [#2723 (comment)](https://github.com/aws/karpenter/issues/2723#issuecomment-1301371762) be a part of this PR or some follow-up work ?

That would be a follow-up at some point. We have some ideas on how to do that, but we don't have a timeline yet.

**Comment by bwagner5** [comment]
I went to test this PR and I'm a little confused why this is needed now. I launched both an i3.16xlarge (8 instance-store disks) and c5ad.xlarge (1 instance-store disk). Both instances had all disk devices when executing `lsblk` without any blockDeviceMappings specified.

```
> lsblk
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
xvda    202:0    0   20G  0 disk
└─xvda1 202:1    0   20G  0 part /
nvme1n1 259:0    0  1.7T  0 disk
nvme0n1 259:1    0  1.7T  0 disk
nvme3n1 259:2    0  1.7T  0 disk
nvme4n1 259:3    0  1.7T  0 disk
nvme2n1 259:4    0  1.7T  0 disk
nvme6n1 259:5    0  1.7T  0 disk
nvme7n1 259:6    0  1.7T  0 disk
nvme5n1 259:7    0  1.7T  0 disk
```

When would virtualName be required?

**Comment by bwagner5** [comment]
Until we can nail down the use-case for this change, I'll close for now.  If there is a use-case that I'm missing, feel free to reopen or discuss in the issue :)

**Comment by sidewinder12s** [comment]
In our environment, we will take any detected instance store volumes and mount and format them in raid0 and then move the kubelet and docker directories onto that device so that pulled images and any containers use it by default (all within the userdata script).

We don't currently expose that fact to users as a resource they request, it's just there to improve performance for users who land on it.

We figure we're paying for it, we should utilize it. I do see there are other projects around for actually making nvme more of a schedulable resource like a device driver, but just utilizing the disks would be a good start even if they are not exposed to Karpenters scheduling logic.

**Comment by cep21** [comment]
@sidewinder12s are you able to share your script or configuration?

**Comment by sidewinder12s** [comment]
We've actually moved to using the build in support for this in the EKS AMI bootstrap flag (still waiting on proper Karpenter support)

**Comment by phyyou** [comment]
> @sidewinder12s are you able to share your script or configuration?

I found similar script in [aws emr containers best practices documentation](https://aws.github.io/aws-emr-containers-best-practices/storage/docs/spark/instance-store/#mount-kubelet-pod-directory-on-nvme-disks)

**Comment by phyyou** [comment]
So I using below script finally.

```
      blockDeviceMappings: []
      userData: |
        MIME-Version: 1.0
        Content-Type: multipart/mixed; boundary="BOUNDARY"

        --BOUNDARY
        Content-Type: text/x-shellscript; charset="us-ascii"

        #!/bin/bash
        echo "Running a custom user data script"
        set -ex

        yum install -y mdadm nvme-cli
        nvme_disks=($(nvme list | grep "Amazon EC2 NVMe Instance Storage" | awk -F'[[:space:]][[:space:]]+' '{print $1}')) && [[ $${#nvme_disks[@]} -eq 1 ]] && mkfs.xfs -F $${nvme_disks[*]} && systemctl stop docker && mkdir -p /var/lib/kubelet/pods && mount $${nvme_disks[*]} /var/lib/kubelet/pods && chmod 750 /var/lib/docker && systemctl start docker
        nvme_disks=($(nvme list | grep "Amazon EC2 NVMe Instance Storage" | awk -F'[[:space:]][[:space:]]+' '{print $1}')) && [[ $${#nvme_disks[@]} -ge 2 ]] && mdadm --create --verbose /dev/md0 --level=0 --raid-devices=$${#nvme_disks[@]} $${nvme_disks[*]} && mkfs.xfs -F /dev/md0 && systemctl stop docker && mkdir -p /var/lib/kubelet/pods && mount /dev/md0 /var/lib/kubelet/pods && chmod 750 /var/lib/docker && systemctl start docker
        --BOUNDARY--

**Pull request closed without merging**